### PR TITLE
[PATCH v8] linux-dpdk: pktio: add pci ports using config file

### PR DESCRIPTION
**Pull request by vijayinavolu**
Implementation to support pci ports addition using the odp config file
for linux dpdk

Signed-off-by: Vijay Ram Inavolu <vinavolu@marvell.com>

### CODE REVIEW

File: platform/linux-dpdk/odp_init.c [hunk 1]
```diff
@@ -78,13 +78,79 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+	char *buf = NULL;
+	int pci_count;
+	int i, j;
+
+	for (i = 0; i < 2; i++) {
+		pci_count = _odp_libconfig_lookup_array_str_size(pci_str[i]);
+		if (pci_count == 0)
+			continue;
+
+		pci_list = malloc(pci_count * sizeof(char *));
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:95]: Variable-length array could be used here to simplify the code.

File: platform/linux-dpdk/odp_init.c [hunk 2]
```diff
@@ -78,13 +78,79 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+		if (pci_list == NULL)
+			goto cleanup;
+
+		for (j = 0; j < pci_count; j++) {
+			pci_list[j] = malloc(sizeof(char) * 100);
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:100]: Please replace the value `100`, which is used several time in this function,  with either a variable or define.

File: platform/linux-generic/include/odp_libconfig_internal.h [hunk 3]
```diff
@@ -22,6 +22,8 @@ int _odp_libconfig_term_global(void);
 
 int _odp_libconfig_lookup_int(const char *path, int *value);
 int _odp_libconfig_lookup_array(const char *path, int value[], int max_num);
+int _odp_libconfig_lookup_array_str(const char *path, char **value,
+				    int max_count, unsigned int max_str);
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/include/odp_libconfig_internal.h:26]: Done

File: platform/linux-generic/odp_libconfig.c [hunk 4]
```diff
@@ -111,6 +111,89 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
 	return  (ret_def == CONFIG_TRUE || ret_rt == CONFIG_TRUE) ? 1 : 0;
 }
 
+int _odp_libconfig_lookup_array_str_size(const char *path)
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: submitted licconfig changes are separate commit here

File: platform/linux-generic/odp_libconfig.c [hunk 5]
```diff
@@ -111,6 +112,61 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+		/* array is empty, check the next config */
+		if (num == 0)
+			continue;
+
+		/* return the size of array */
+		if (max_count == 0)
+			return num;
+
+		if (num > max_count)
+			return 0;
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: done

File: platform/linux-generic/odp_libconfig.c [hunk 6]
```diff
@@ -111,6 +112,61 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+			config = &odp_global_ro.libconfig_default;
+
+		setting = config_lookup(config, path);
+
+		if (setting == NULL)
+			continue;
+
+		/* invalid config if element is not an array */
+		if (config_setting_is_array(setting) == CONFIG_FALSE)
+			return 0;
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: return 0 is following the other functions in current libconfig, but returning error is better. Changed all error conditions to return negative now.

File: platform/linux-generic/odp_libconfig.c [hunk 7]
```diff
@@ -111,6 +112,61 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+
+		num = config_setting_length(setting);
+
+		/* array is empty, check the next config */
+		if (num == 0)
+			continue;
+
+		/* return the size of array */
+		if (max_count == 0)
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: combined max_count 0 and value NULL to one.

File: platform/linux-generic/odp_libconfig.c [hunk 8]
```diff
@@ -111,6 +112,61 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+			return num;
+
+		if (num > max_count)
+			return 0;
+
+		elem = config_setting_get_elem(setting, 0);
+		if (config_setting_type(elem) != CONFIG_TYPE_STRING)
+			return 0;
+
+		if (value == NULL)
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: done

File: platform/linux-generic/odp_libconfig.c [hunk 9]
```diff
@@ -111,6 +112,61 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+			elem = config_setting_get_elem(setting, j);
+			if (strlen(elem->value.sval) > str_size)
+				return j;
+			strcpy(value[j], elem->value.sval);
+		}
+
+		return num;
+	}
+
+	return 0;
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: Updated based on the new error return value by the libconfig function

File: config/odp-linux-dpdk.conf [hunk 10]
```diff
@@ -45,6 +45,23 @@ dpdk: {
 	# should reserve all shared resources and configure the system before
 	# forking child processes for the best success probability.
 	process_mode_memory_mb = 512
+
+	# PCI options
+	pci: {
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply config/odp-linux-dpdk.conf]: updated in the used files

File: platform/linux-dpdk/odp_init.c [hunk 11]
```diff
@@ -78,13 +78,61 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+	const int str_size =  100;
+	char *buf = NULL;
+	int pci_count;
+	int i, j;
+
+	for (i = 0; i < 2; i++) {
+		/* get the size of the array */
+		pci_count = _odp_libconfig_lookup_array_str(pci_str[i], NULL, 0, 0);
+		if (pci_count == 0)
+			continue;
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: Updated based on the new error return value by the libconfig functio

File: platform/linux-dpdk/odp_init.c [hunk 12]
```diff
@@ -78,13 +78,61 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+		char *pci_list_addr[pci_count];
+
+		for (j = 0; j < pci_count; j++)
+			pci_list_addr[j] = pci_list[j];
+
+		if (pci_count != _odp_libconfig_lookup_array_str(pci_str[i],
+								 pci_list_addr, pci_count, str_size))
+			return 0;
+
+		buf = malloc(pci_count * str_size);
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: Size of 100 should be enough for pci address and -w/-b but added extra 3 byte allocation for -, w/b and a space for code clarity

File: platform/linux-generic/odp_libconfig.c [hunk 13]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+ * Returns the number of strings in an string array setting. Outputs up
+ * to 'count' strings when the 'value' array pointer is not NULL. If return
+ * value is larger than 'count', there are more strings than the function was
+ * allowed to output. If return value (N) is less than 'count', only
+ * strings[0 ... N-1] have been written.
+ *
+ * @param          Path of the setting
+ * @value[out]     Array of strings to be copied from the setting
+ * @count          Number of strings to be copied
+ * @str_size       Maxiumim string allowed to be copied
```
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:127]: We generally use the following argument documentation style:
```
 * @param      path   Path of the setting
 * @param[out] value  Array of strings to be copied from the setting
```

File: platform/linux-generic/odp_libconfig.c [hunk 14]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+			return -1;
+		}
+		num = config_setting_length(setting);
+
+		/* array is empty, check the next config */
+		if (num == 0)
+			continue;
+
+		/* return the size of array */
+		if (count == 0 || value == NULL)
```
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:162]: Should be `max_count`

File: platform/linux-generic/odp_libconfig.c (hunk 13 above)
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:127]: The variable names mentioned in the doc should match the argument names.

File: platform/linux-generic/odp_libconfig.c [hunk 15]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+			ODP_ERR("libconfig: %s array is not of type string\n", path);
+			return -1;
+		}
+
+		for (j = 0; j < num; j++) {
+			elem = config_setting_get_elem(setting, j);
+			if (strlen(elem->value.sval) > str_size) {
+				ODP_ERR("libconfig: length of %s bigger than size %d\n",
+					elem->value.sval,
+					     strlen(elem->value.sval), str_size);
```
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:176]: Invalid indentation.

File: platform/linux-generic/odp_libconfig.c [hunk 16]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+				return -1;
+			}
+			strcpy(value[j], elem->value.sval);
+		}
+
+		return num;
+	}
+
+	return -1;
```
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:185]: Please add error print also here.

File: platform/linux-generic/odp_libconfig.c [hunk 17]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+		/* invalid config if element is not an array */
+		if (config_setting_is_array(setting) == CONFIG_FALSE) {
+			ODP_ERR("libconfig: %s is not an array\n", path);
+			return -1;
+		}
+		num = config_setting_length(setting);
+
+		/* array is empty, check the next config */
+		if (num == 0)
+			continue;
```
Comment (MatiasElo) [inline platform/linux-generic/odp_libconfig.c:159]: The first found setting should be selected (e.g. runtime overrides default), so zero should be returned in this case.

File: platform/linux-dpdk/odp_init.c [hunk 18]
```diff
@@ -78,13 +78,63 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+	char *buf = NULL;
+	int pci_count;
+	int i, j;
+
+	for (i = 0; i < 2; i++) {
+		/* get the size of the array */
+		pci_count = _odp_libconfig_lookup_array_str(pci_str[i], NULL, 0, 0);
+
+		/* skip if list is emply or any config error */
+		if (pci_count <= 0)
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:95]: If there was an error in parsing the config file it's safer to just fail.

File: platform/linux-dpdk/odp_init.c [hunk 19]
```diff
@@ -78,13 +78,63 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+		char pci_list[pci_count][str_size];
+		char *pci_list_addr[pci_count];
+
+		for (j = 0; j < pci_count; j++)
+			pci_list_addr[j] = pci_list[j];
+
+		if (pci_count != _odp_libconfig_lookup_array_str(pci_str[i],
+								 pci_list_addr, pci_count, str_size))
+			return 0;
+		/* Buffer to concatinate list of '-w/-b <pci addr>' strings */
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:107]: Typo: `concatenate`

File: platform/linux-dpdk/odp_init.c [hunk 20]
```diff
@@ -78,13 +78,63 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+			continue;
+
+		char pci_list[pci_count][str_size];
+		char *pci_list_addr[pci_count];
+
+		for (j = 0; j < pci_count; j++)
+			pci_list_addr[j] = pci_list[j];
+
+		if (pci_count != _odp_libconfig_lookup_array_str(pci_str[i],
+								 pci_list_addr, pci_count, str_size))
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:105]: Line over 100 characters.

File: platform/linux-dpdk/odp_init.c [hunk 21]
```diff
@@ -78,13 +78,63 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+			return 0;
+		/* Buffer to concatinate list of '-w/-b <pci addr>' strings */
+		buf = malloc(pci_count * (str_size+3));
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:108]: Spaces preferred around '+'

File: platform/linux-dpdk/odp_init.c [hunk 22]
```diff
@@ -95,25 +145,32 @@ static int _odp_init_dpdk(const char *cmdline)
...
+	ODP_PRINT("  %s: %" PRIu32 "\n", str, mem_prealloc);
 
 	if (cmdline == NULL) {
 		cmdline = getenv("ODP_PLATFORM_PARAMS");
 		if (cmdline == NULL)
 			cmdline = "";
 	}
 
-	cmdlen = snprintf(NULL, 0, "odpdpdk -m %" PRIu32 " %s ", mem_prealloc, cmdline);
+	pcicmdlen = read_pci_config(&pci_cmd);
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:156]: As mentioned above, it would be safer to fail if reading the config file failed.

**Review by MatiasElo — COMMENTED** [review]
The order of the commits should still be reversed.

File: platform/linux-generic/odp_libconfig.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: changed

File: platform/linux-generic/odp_libconfig.c [hunk 23]
```diff
@@ -111,6 +112,79 @@ int _odp_libconfig_lookup_int(const char *path, int *value)
...
+ * allowed to output. If return value (N) is less than 'count', only
+ * strings[0 ... N-1] have been written.
+ *
+ * @param          Path of the setting
+ * @value[out]     Array of strings to be copied from the setting
+ * @count          Number of strings to be copied
+ * @str_size       Maxiumim string allowed to be copied
+ *
+ * @return Number of strings in the setting
+ * @retval <0 on failureodp_
```
[the remark this answers is not in the document]
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: corrected

File: platform/linux-dpdk/odp_init.c (hunk 22 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c:160]: changed

File: platform/linux-generic/odp_libconfig.c (hunk 16 above)
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c:181]: added

File: platform/linux-generic/odp_libconfig.c (hunk 14 above)
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: changed

File: platform/linux-generic/odp_libconfig.c (hunk 13 above)
Comment (vijayinavolu) [reply platform/linux-generic/odp_libconfig.c]: done

File: platform/linux-dpdk/odp_init.c (hunk 18 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: done

File: platform/linux-dpdk/odp_init.c (hunk 20 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: changed

File: platform/linux-dpdk/odp_init.c (hunk 21 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: added

File: platform/linux-dpdk/odp_init.c (hunk 22 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c:160]: done

File: platform/linux-dpdk/odp_init.c [hunk 24]
```diff
@@ -78,13 +78,67 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+		char pci_list[pci_count][str_size];
+		char *pci_list_addr[pci_count];
+
+		for (j = 0; j < pci_count; j++)
+			pci_list_addr[j] = pci_list[j];
+
+		if (pci_count != _odp_libconfig_lookup_array_str(pci_str[i],
+								 pci_list_addr,
+								 pci_count, str_size))
+			return 0;
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:110]: Should return failure.

File: platform/linux-dpdk/odp_init.c [hunk 25]
```diff
@@ -78,13 +78,67 @@ static void disable_features(odp_global_data_ro_t *global_ro,
...
+		/* Buffer to concatenate list of '-w/-b <pci addr>' strings */
+		buf = malloc(pci_count * (str_size + 3));
+		if (buf == NULL)
+			return 0;
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_init.c:114]: Please add error print and failure return code here.

**Review by MatiasElo — COMMENTED** [review]
Two minor comments but otherwise the PR is looking good. The conf file documentation has also some typos (e.g. missing periods).

The commit messages are missing capital letters and periods. You can add my reviewed-by tag in the next version.

File: platform/linux-dpdk/odp_init.c (hunk 24 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: changed

File: platform/linux-dpdk/odp_init.c (hunk 25 above)
Comment (vijayinavolu) [reply platform/linux-dpdk/odp_init.c]: added

**Review by MatiasElo — APPROVED** [review]

**Pull request merged**